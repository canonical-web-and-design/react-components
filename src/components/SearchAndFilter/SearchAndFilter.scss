@import "~vanilla-framework/scss/base_placeholders";

.search-and-filter {
  span[class*="p-contextual-menu"],
  .p-contextual-menu__dropdown {
    display: block;
    max-width: 100%;
    width: 100%;
  }

  .p-contextual-menu__dropdown {
    padding: 0.5rem;
  }

  .p-search-box {
    margin: 0;
  }

  .p-contextual-menu__dropdown {
    padding: 1rem;
    top: 0;
    left: 0;
  }

  &__panel {
    padding: 0;
    transition: opacity 0.25s;

    &[aria-hidden="true"] {
      opacity: 0;
      pointer-events: none;
    }
  }

  &__section {
    border-bottom: 1px solid $color-mid-x-light;
    padding-bottom: 1rem;

    &:last-child {
      border: 0;
    }
  }

  &__section-header {
    color: $color-mid-dark;
    font-size: 1rem;
    margin-bottom: 0;
  }
}
