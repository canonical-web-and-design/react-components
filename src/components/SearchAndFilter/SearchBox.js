import classNames from "classnames";
import React from "react";
import PropTypes from "prop-types";

const SearchBox = React.forwardRef(
  (
    {
      autocomplete = "off",
      className,
      disabled,
      onChange,
      onSubmit,
      placeholder = "Search",
      value,
      ...props
    },
    ref
  ) => {
    const submit = (e) => {
      e.preventDefault();
      onSubmit && onSubmit(e);
    };
    return (
      <form
        className={classNames("p-search-and-filter__box", className)}
        onSubmit={submit}
        {...props}
        ref={ref}
      >
        <label className="u-off-screen" htmlFor="search">
          {placeholder || "Search"}
        </label>
        <input
          autoComplete={autocomplete}
          className="p-search-and-filter__input"
          disabled={disabled}
          id="search"
          name="search"
          onChange={(evt) => onChange(evt.target.value)}
          placeholder={placeholder}
          type="search"
          value={value}
        />
        <button
          alt="search"
          className="u-off-screen"
          disabled={disabled}
          type="submit"
        >
          Search
        </button>
      </form>
    );
  }
);

SearchBox.propTypes = {
  autocomplete: PropTypes.string,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

SearchBox.displayName = "SearchBox";

export default SearchBox;
