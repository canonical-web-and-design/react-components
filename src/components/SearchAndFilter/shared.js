/**
 * Return number of overflowing chips given a row threshold
 * @param {array} chips - An array of chips
 * @param {Integer} overflowRowLimit - Number of rows to show before counting
 * overflow
 */
export const overflowingChipsCount = (chips, overflowRowLimit) => {
  let overflowChips = 0;
  if (chips) {
    chips.forEach((chip) => {
      if (chip.offsetTop > chip.offsetHeight * overflowRowLimit)
        overflowChips++;
    });
  }
  return overflowChips;
};
