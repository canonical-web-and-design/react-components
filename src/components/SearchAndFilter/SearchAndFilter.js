import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";

import SearchBox from "../SearchBox";
import ContextualMenu from "../ContextualMenu";
import FilterPanelSection from "./FilterPanelSection";
import Chip from "../Chip";
import { overflowingChipsCount } from "./shared";

import "./SearchAndFilter.scss";

const SearchAndFilter = ({
  externallyControlled = false,
  onChange,
  filterPanelData,
}) => {
  const [searchData, setSearchData] = useState([]);
  const [searchTerm, setSearchTerm] = useState("");
  const [filterPanelHidden, setFilterPanelHidden] = useState(true);
  const [searchBoxExpanded, setSearchBoxExpanded] = useState(false);
  const [overflowSearchTermCounter, setOverflowSearchTermCounter] = useState(0);
  const filterPanelRef = useRef();
  const searchContainerRef = useRef();

  const searchOnChange = (searchTerm) => {
    setSearchTerm(searchTerm);
    if (externallyControlled && onChange) {
      onChange(searchTerm);
    }
  };

  // This useEffect sets up listeners so the panel will close if user clicks
  // anywhere else on the page or hits the escape key
  useEffect(() => {
    const closePanel = () => {
      setFilterPanelHidden(true);
    };

    const mouseDown = (e) => {
      // Check if click is outside of filter panel
      if (!filterPanelRef?.current?.contains(e.target)) {
        // If so, close the panel
        closePanel();
      }
    };

    const keyDown = (e) => {
      if (e.code === "Escape") {
        // Close panel if Esc keydown detected
        closePanel();
      }
    };

    // Add listener on document to capture click events
    document.addEventListener("mousedown", mouseDown);
    // Add listener on document to capture keydown events
    document.addEventListener("keydown", keyDown);
    // return function to be called when unmounted
    return () => {
      document.removeEventListener("mousedown", mouseDown);
      document.removeEventListener("keydown", keyDown);
    };
  }, [searchTerm]);

  // Add passed chip to the searchData array
  const addToSelected = (chip) => {
    const currentSelected = [...searchData];
    if (!currentSelected.includes(chip)) {
      currentSelected.push(chip);
      setSearchData(currentSelected);
    }
  };

  // Remove passed chip from the searchData array
  const removeFromSelected = (chip) => {
    if (searchData.includes(chip)) {
      const updatedSelected = searchData.filter(
        (searchDataChip) => searchDataChip !== chip
      );
      setSearchData(updatedSelected);
    }
  };

  const handleSubmit = (e) => {
    if (searchTerm !== "") {
      addToSelected({ value: searchTerm });
      setSearchTerm("");
    }
  };

  // If the offsetTop is more than double height of a single chip, consider it
  // overflowing
  const updateFlowCount = function () {
    const chips = searchContainerRef?.current?.querySelectorAll(".p-chip");
    const overflowCount = overflowingChipsCount(chips, 1);
    setOverflowSearchTermCounter(overflowCount);
  };

  useEffect(() => {
    if (typeof ResizeObserver !== "undefined") {
      const wrapperWidthObserver = new ResizeObserver(() => {
        updateFlowCount();
      });
      const wrapper = searchContainerRef.current;
      wrapperWidthObserver.observe(wrapper);
    } else {
      updateFlowCount();
    }
  }, [searchData]);

  return (
    <div className="search-and-filter" ref={filterPanelRef}>
      <div
        className="search-and-filter__search-container"
        aria-expanded={searchBoxExpanded}
        ref={searchContainerRef}
      >
        {Object.values(searchData).map((chip) => {
          return (
            <Chip
              lead={chip.lead}
              value={chip.value}
              key={`search-${chip.lead}+${chip.value}`}
              onDismiss={() => removeFromSelected(chip)}
              selected={true}
            />
          );
        })}
        <SearchBox
          autocomplete="off"
          externallyControlled={externallyControlled}
          placeholder={searchData.length ? "Add filter" : "Search and filter"}
          onChange={(searchTerm) => searchOnChange(searchTerm)}
          onClick={() => setFilterPanelHidden(false)}
          onFocus={() => setFilterPanelHidden(false)}
          onSubmit={(e) => handleSubmit(e)}
          value={searchTerm}
        />
        {overflowSearchTermCounter > 0 && (
          <span
            className="search-and-filter__selected-count"
            onClick={() => setSearchBoxExpanded(true)}
            onKeyDown={() => setSearchBoxExpanded(true)}
            role="button"
          >
            +{overflowSearchTermCounter}
          </span>
        )}
      </div>
      {filterPanelData && (
        <div
          className="search-and-filter__panel"
          aria-hidden={filterPanelHidden}
        >
          <ContextualMenu>
            {filterPanelData.map((filterPanelSectionData) => {
              return (
                <div key={filterPanelSectionData.id}>
                  <FilterPanelSection
                    data={filterPanelSectionData}
                    addToSelected={addToSelected}
                    searchData={searchData}
                  />
                </div>
              );
            })}
          </ContextualMenu>
        </div>
      )}
    </div>
  );
};

SearchAndFilter.propTypes = {
  externallyControlled: PropTypes.bool,
  searchData: PropTypes.arrayOf(
    PropTypes.shape({
      chips: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          lead: PropTypes.string,
          value: PropTypes.string,
        })
      ),
    })
  ),
  filterPanelData: PropTypes.arrayOf(
    PropTypes.shape({
      heading: PropTypes.string,
      chips: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number,
          lead: PropTypes.string,
          value: PropTypes.string,
        })
      ),
    })
  ),
  onChange: PropTypes.func,
  onSubmit: PropTypes.func,
};

export default SearchAndFilter;
