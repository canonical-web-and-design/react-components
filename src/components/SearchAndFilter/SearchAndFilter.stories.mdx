import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { withKnobs } from "@storybook/addon-knobs";
import SearchAndFilter from "./SearchAndFilter";

<Meta
  title="SearchAndFilter"
  component={SearchAndFilter}
  decorators={[withKnobs]}
/>

### Search And Filter

This component is used to search and filter large datasets.

### Props

<Props of={SearchAndFilter} />

### Default

<Preview>
  <Story name="Default">
    <SearchAndFilter />
  </Story>
</Preview>

### External state

If you wish to control the value of the data set via external state you can set
the `externallyControlled` prop and provide an `onChange` method to update the
state and the `value` from state.

<Preview>
  <Story name="External state">
    <SearchAndFilter
      externallyControlled
      onChange={(searchTerm) => {
        console.log("onChange: ", searchTerm);
      }}
      filterPanelData={[
        {
          id: 0,
          heading: "Cloud",
          chips: [
            { lead: "Cloud", value: "Google" },
            { lead: "Cloud", value: "AWS" },
            { lead: "Cloud", value: "Azure" },
          ],
        },
        {
          id: 1,
          heading: "Region",
          chips: [
            { lead: "Region", value: "us-east1" },
            { lead: "Region", value: "us-north2" },
            { lead: "Region", value: "us-south3" },
            { lead: "Region", value: "us-north4" },
            { lead: "Region", value: "us-east5" },
            { lead: "Region", value: "us-south6" },
            { lead: "Region", value: "us-east7" },
            { lead: "Region", value: "us-east8" },
            { lead: "Region", value: "us-east9" },
            { lead: "Region", value: "us-east10" },
          ],
        },
        {
          id: 2,
          heading: "Owner",
          chips: [
            { lead: "Owner", value: "foo" },
            { lead: "Owner", value: "bar" },
            { lead: "Owner", value: "baz" },
          ],
        },
      ]}
    />
  </Story>
</Preview>
