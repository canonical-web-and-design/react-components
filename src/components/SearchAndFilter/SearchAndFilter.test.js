import { shallow, mount } from "enzyme";
import React from "react";

import SearchAndFilter from "./SearchAndFilter";

describe("Search and filter", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchAndFilter />);
    expect(wrapper).toMatchSnapshot();
  });

  it("hide the clear button when there is no value in search box", () => {
    const wrapper = mount(<SearchAndFilter />);
    expect(wrapper.find(".p-search-box__reset").exists()).toBe(false);
  });

  it("onChange handler called", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter externallyControlled onChange={mockOnChange} />
    );
    wrapper.find(".p-search-box__input").simulate("change");
    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });

  it("shows panel on focus", () => {
    const mockOnChange = jest.fn();
    const wrapper = mount(
      <SearchAndFilter externallyControlled onChange={mockOnChange} />
    );
    expect(
      wrapper.find(".search-and-filter__panel").prop("data-visible")
    ).toEqual(false);
    wrapper.find(".p-search-box__input").simulate("focus");
    expect(
      wrapper.find(".search-and-filter__panel").prop("data-visible")
    ).toEqual(true);
  });
});
