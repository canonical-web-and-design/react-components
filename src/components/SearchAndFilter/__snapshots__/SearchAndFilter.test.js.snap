// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search and filter renders 1`] = `
<div
  className="search-and-filter"
  onClick={[Function]}
>
  <div
    aria-expanded={false}
    className="search-and-filter__search-container"
    data-active={true}
    data-empty={true}
  >
    <SearchBox
      autocomplete="off"
      data-overflowing={false}
      externallyControlled={true}
      onChange={[Function]}
      onSubmit={[Function]}
      placeholder="Search and filter"
      value=""
    />
  </div>
</div>
`;
